Build/Test Tools: Comment on a PR when no Trac ticket is included

anamarijapapic opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

When contributing to Core using a pull request on GitHub, a [Trac ticket is required for the contribution to be considered](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/).

Even though the pull request template notes this, it's easy to forget or overlook.

Adding a job in the Pull Request Comments workflow to add a reminder comment when a Trac ticket link is missing is an easy way to help avoid contributions being lost in the shuffle.

Trac ticket: https://core.trac.wordpress.org/ticket/60129

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

anamarijapapic committed:
Build/Test Tools: Comment on a PR when no Trac ticket is included

Add a job in the Pull Request Comments workflow to add a reminder comment when a Trac ticket link is missing.

See Trac ticket #60129.

desrosj commented on .github/workflows/pull-request-comments.yml:
```diff
@@ -163,3 +163,46 @@ jobs:
+            const prBody = pr.data.body ?? '';
+
+            const tracTicketRegex = new RegExp( 'https?://core.trac.wordpress.org/ticket/([0-9]+)', 'g' );
+            const tracTicketMatches = prBody.match( tracTicketRegex );
```
It's not obvious (or recommended), but it is possible to also include the Trac ticket URL in the PR title. I tested this to confirm in #6088. You'll notice it's listed on the Trac ticket. We should check for it just in case, but let's not recommend this practice.

desrosj commented on .github/workflows/pull-request-comments.yml:
```diff
@@ -163,3 +163,46 @@ jobs:
+
+            To attach a pull request to a Trac ticket, please include the ticket's full URL in your pull request description.
+
+            More information about how GitHub pull requests can be used to contribute to WordPress can be found in [this blog post](https://make.wordpress.org/core/2020/02/21/working-on-trac-tickets-using-github-pull-requests/).
```
I know that this is copied form the welcome message above. But let's change it a bit. There is now a [page in the handbook that has been added to document this](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/). Let's update that link here and in the welcome message above. Instead of "this blog post", let's just say "the Core Handbook".

desrosj commented on .github/workflows/pull-request-comments.yml:
```diff
@@ -163,3 +163,46 @@ jobs:
+                repo: context.repo.repo,
+                issue_number: context.issue.number,
+                body: `## Trac Ticket Missing
+            Hi @${{ github.actor }}! 👋
```
This PR made it clear that we have a number of bot-powered comments now and we need to be mindful of that when introducing new ones. 

I think it's important for the welcome comment to be more friendly and welcoming. But this one can probably be more direct. Let's remove the greeting, and change the language a bit.

"This pull request is missing a Trac ticket link. For a contribution to be considered, there must be a corresponding ticket in Trac.

To attach a pull request to a Trac ticket, please include the ticket's full URL in your pull request description. More information about contributing to WordPress on GitHub can be found in the Core Handbook".

desrosj reviewed (changes requested):
Hi @anamarijapapic!

Thanks for this PR. I think it is looking really good! I left a few requests to help tighten this comment up a bit.

peterwilsoncc commented on .github/workflows/pull-request-comments.yml:
```diff
@@ -163,3 +163,46 @@ jobs:
+    permissions:
+      issues: write
+      pull-requests: write
+    if: ${{ github.repository == 'WordPress/wordpress-develop' && github.event_name == 'pull_request_target' }}
```
```suggestion
    if: ${{ github.repository == 'WordPress/wordpress-develop' && github.event_name == 'pull_request_target' && github.event.pull_request.draft == false }}
```

It would be good to wait until a PR is out of draft before adding this comment. 

When starting on a ticket, it's sometimes helpful to leave out the link to the ticket intentionally so the PR doesn't show up on trac until the code is ready to be reviewed.

desrosj replied:
Good call, @peterwilsoncc! There's some tested logic around this in the Props Bot workflow we could borrow here.

desrosj commented on .github/workflows/pull-request-comments.yml:
```diff
@@ -163,3 +163,46 @@ jobs:
+    permissions:
+      issues: write
+      pull-requests: write
+    if: ${{ github.repository == 'WordPress/wordpress-develop' && github.event_name == 'pull_request_target' }}
```
```suggestion
    if: ${{ github.repository == 'WordPress/wordpress-develop' && github.event_name == 'pull_request_target' && ! github.event.pull_request.draft && github.event.pull_request.state == 'open' }}
```

We should also make sure the PR is still open to prevent comments on any that have been closed. I also switched the other suggestion to a `!` comparison instead.

desrosj commented:
Was thinking about this a bit more today. I think we should also attempt to check if a previous comment exists to avoid multiple comments for the same thing on a PR.

anamarijapapic commented:
Thank you for the review @desrosj and @peterwilsoncc!

I'll make the requested changes in the next few days when I have some free time. Feel free to let me know if you have any further suggestions!

anamarijapapic committed:
Check that a PR is out of draft and is still open

- Wait until a PR is out of draft before adding 'Trac Ticket Missing' comment
- Make sure the PR is still open to prevent adding 'Trac Ticket Missing' comment on closed PRs

Co-authored-by: Jonathan Desrosiers <359867+desrosj@users.noreply.github.com>

anamarijapapic committed:
Check for Trac ticket URL in the PR title

anamarijapapic committed:
Link to the Core Handbook in the 'Welcome' and 'Trac Ticket Missing' comments

- Instead of a blog post, there is now a page in the handbook that has been added to document GitHub pull requests x Trac tickets

anamarijapapic committed:
Improve the 'Trac Ticket Missing' comment message

- Remove the greeting
- Change the language a bit

anamarijapapic committed:
Check if a previous comment exists.

- Check if a previous 'Trac Ticket Missing' comment exists to avoid multiple comments for the same thing on a PR

anamarijapapic committed:
Optimize script for trac-ticket-check

- Utilize destructuring and reduce the number of unnecessary property accesses
- Optimize looping through all comments

desrosj commented:
Merged in https://core.trac.wordpress.org/changeset/58092.
